### PR TITLE
Update split_non_commuting transform to accept any type of measurement (Issue #4932)

### PR DESCRIPTION
**Pull request by abhishekabhishek**
**Context:**

Update the `split_non_commuting` transform to accept measurements which accept both wires and observables i.e. `probs`, `sample` and `counts`

**Description of the Change:**

1. Modified the `split_non_commuting` transform in [pennylane/transforms/split_non_commuting.py](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/transforms/split_non_commuting.py) by adding a condition for `probs`, `sample` and `counts` and handling `obs.wires` as a `qml.PauliZ(wires[0])@qml.PauliZ(wires[1])@...@qml.PauliZ(wires[len(wires)-1])` observable.
2. Updated the creation of new tapes `split_non_commuting` to handle measurements initialized using wires
3. Added an entry under Improvements/Community contributions in `changelog-dev.md`
4. Added unit, integration and automatic differentiation tests for all interfaces in `tests/transforms/split_non_commuting.py`
5. Modified existing integration and automatic differentiation tests to explicitly apply the transform on QNodes since it was previously applied under the hood

**Benefits:**

Allows to split tapes into commuting groups for measurements initialized using both wires and observables.

**Related GitHub Issues:**
#4932

### CODE REVIEW

**Comment by rmoyard** [comment]
Hello @abhishekabhishek! Thanks for opening the PR, let me know when it is ready for review and I will take a look.

**Comment by abhishekabhishek** [comment]
> Hello @abhishekabhishek! Thanks for opening the PR, let me know when it is ready for review and I will take a look.

Hi @rmoyard, the PR is now ready for review. Please feel free to have a look and let me know if I need to make any changes.

File: tests/transforms/test_split_non_commuting.py [hunk 1]
```diff
@@ -13,105 +13,206 @@
...
+        with qml.queuing.AnnotatedQueue() as q:
+            qml.PauliZ(0)
+            qml.Hadamard(0)
+            qml.CNOT((0, 1))
+            for obs in obs_list:
+                meas_type(op=obs)
+
+        tape = qml.tape.QuantumScript.from_queue(q)
+        _, fn = split_non_commuting(tape)
+        assert all(np.array(fn(group_coeffs)) == np.arange(len(tape.measurements)))
```
Comment (rmoyard) [inline tests/transforms/test_split_non_commuting.py:165]: For `obs_list_2` and `obs_list_3`, they are as many groups as measurements?

**Review by rmoyard — COMMENTED** [review]
Thanks for your PR @abhishekabhishek ! I have left an initial set of comments.

File: pennylane/transforms/split_non_commuting.py [hunk 2]
```diff
@@ -133,33 +140,62 @@ def circuit0(x):
...
+        >>> [t.measurements for t in tapes]
+        [[expval(PauliX(wires=[0])), probs(wires=[1])], [probs(wires=[0, 1])]]
+    """
 
-    obs_list = tape.observables
+    # Construct a list of observables to group based on the measurements in the tape
+    obs_list = []
+    for obs in tape.observables:
+        # observable provided for a measurement
+        if isinstance(obs, qml.operation.Observable):
```
[the remark this answers is not in the document]
Comment (abhishekabhishek) [reply pennylane/transforms/split_non_commuting.py:178]: For measurements initialized using an observable, we have an observable here. For measurements initialized using wires, we have a measurement process here. For e.g.
```
with qml.queuing.AnnotatedQueue() as q:
    qml.expval(qml.PauliZ(0))
    qml.var(qml.PauliZ(0))
    qml.probs([0, 1])
    qml.sample(wires=[0, 1])
    qml.sample(op=qml.PauliX(0))

tape = qml.tape.QuantumScript.from_queue(q)

>>> tape.observables
[PauliZ(wires=[0]),
 PauliZ(wires=[0]),
 probs(wires=[0, 1]),
 sample(wires=[0, 1]),
 PauliX(wires=[0])]
```

File: pennylane/transforms/split_non_commuting.py [hunk 3]
```diff
@@ -133,33 +140,62 @@ def circuit0(x):
...
+            obs_list.append(obs)
+        # measurements using wires
+        elif any(isinstance(obs, mp) for mp in [CountsMP, ProbabilityMP, SampleMP]):
```
[the remark this answers is not in the document]
Comment (abhishekabhishek) [reply pennylane/transforms/split_non_commuting.py:181]: Yes, thanks! I'll change this condition statement to:
```
if isinstance(obs, qml.operation.Observable):
    obs_list.append(obs)
else:
    # create the PauliZ tensor product observable when only wires are provided for the measurements
```

File: pennylane/transforms/split_non_commuting.py [hunk 4]
```diff
@@ -133,33 +140,62 @@ def circuit0(x):
...
+            # create the PauliZ tensor product observable when only wires are provided for the
+            # measurements
+            pauliz_obs = qml.PauliZ(obs.wires[0])
+            for wire in obs.wires[1:]:
+                pauliz_obs = pauliz_obs @ qml.PauliZ(wire)
```
[the remark this answers is not in the document]
Comment (abhishekabhishek) [reply pennylane/transforms/split_non_commuting.py:186]: Thanks. I'm trying to use `qml.prod` here as following:
```
pauliz_obs_list = [qml.PauliZ(obs.wires[idx]) for idx in range(len(obs.wires))]
obs_list.append(qml.prod(*pauliz_obs_list))
```
but run into an issue when trying to use `qml.pauli.group_observables`.

```
obs = qml.probs(wires=[0, 1])
pauliz_obs_list = [qml.PauliZ(obs.wires[idx]) for idx in range(len(obs.wires))]
prod_obs = qml.prod(*pauliz_obs_list)

>>> prod_obs
PauliZ(wires=[0]) @ PauliZ(wires=[1])
>>> type(prod_obs)
<class 'pennylane.ops.op_math.prod.Prod'>

obs_list = [prod_obs, qml.PauliX(0)]
group, group_coeffs = qml.pauli.group_observables(obs_list, range(len(obs_list)))

>>> print(group, group_coeffs)
[[Identity(wires=[0]), PauliX(wires=[0])]] [array([1])]
```
whereas the expected output is:
```
obs = qml.probs(wires=[0, 1])
pauliz_obs = qml.PauliZ(obs.wires[0])
for wire in obs.wires[1:]:
    pauliz_obs = pauliz_obs @ qml.PauliZ(wire)
    
>>> pauliz_obs
PauliZ(wires=[0]) @ PauliZ(wires=[1])
>>> type(pauliz_obs)
pennylane.operation.Tensor

obs_list = [pauliz_obs, qml.PauliX(0)]
group, group_coeffs = qml.pauli.group_observables(obs_list, range(len(obs_list)))

>>> print(group, group_coeffs)
[[PauliZ(wires=[0]) @ PauliZ(wires=[1])], [PauliX(wires=[0])]] [array([0]), array([1])]
```

What would be the correct way to use `qml.pauli.group_observables` or is there an alternative method I should be using which would support `qml.prod`?

File: tests/transforms/test_split_non_commuting.py (hunk 1 above)
Comment (abhishekabhishek) [reply tests/transforms/test_split_non_commuting.py:165]: No, there are 2 and 3 groups respectively. This here is to test the `reorder_fn` which ensures that the measurement results are returned in the correct order.

File: pennylane/transforms/split_non_commuting.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rmoyard) [reply pennylane/transforms/split_non_commuting.py:186]: Thanks for trying @abhishekabhishek , apparently the operator arithmetic is not compatible with `group_observables`. Therefore you can keep the implementation with the tensor. Could add a
`# TODO: Revisit when qml.prod is compatible with qml.pauli.group_observables` though?

[the remark this answers is not in the document]
Comment (abhishekabhishek) [reply pennylane/transforms/split_non_commuting.py:186]: Sounds good. I've added the `# TODO`.

File: doc/releases/changelog-dev.md [hunk 5]
```diff
@@ -1,694 +1,25 @@
 :orphan:
 
-# Release 0.34.0-dev (development release)
+# Release 0.35.0-dev (development release)
 
 <h3>New features since last release</h3>
 
-<h4>Statistics and drawings for mid-circuit measurements 🎨</h4>
```
[the remark this answers is not in the document]
Comment (abhishekabhishek) [reply doc/releases/changelog-dev.md]: Yes, the `changelog-dev.md` got updated to `0.35.0`. I've resolved the merge conflicts.

**Review by mudit2812 — APPROVED** [review]
Thanks @abhishekabhishek for the contribution! The implementation is really clean, and the tests are really comprehensive.  Very excited to get this in 😄

File: doc/releases/changelog-dev.md [hunk 6]
```diff
@@ -21,4 +25,4 @@
 
 This release contains contributions from (in alphabetical order):
 
-Matthew Silverman
+Abhishek Abhishek, Matthew Silverman
```
Comment (rmoyard) [inline doc/releases/changelog-dev.md:28]:
```suggestion
Abhishek Abhishek,
Matthew Silverman.
```

**Review by rmoyard — APPROVED** [review]
Thank you @abhishekabhishek, great PR. It looks good on my side!

**Comment by abhishekabhishek** [comment]
> Thanks @abhishekabhishek for the contribution! The implementation is really clean, and the tests are really comprehensive. Very excited to get this in 😄

Thanks @mudit2812 for the review :smile:

**Comment by abhishekabhishek** [comment]
> Thank you @abhishekabhishek, great PR. It looks good on my side!

Thanks @rmoyard for the review and all the help! :smile: :100:

**Pull request merged**